rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+  tally_words: no visible global function definition for ‘tibble’
+  tally_words: no visible global function definition for ‘group_by’
+  tally_words: no visible binding for global variable ‘word’
+  tally_words: no visible global function definition for ‘tally’
+  Undefined global functions or variables:
+    %>% group_by tally tibble word
+```
+
+What we need to do is add these three lines to the bottom of `DESCRIPTION`:
+
+```text
+Imports:
+    dplyr (>= 0.7.0),
+    magrittr (>= 1.5.0)
+```
+
+The `Imports` field in `DESCRIPTION` actually has nothing to do with importing functions;
+it just ensures that those packages are installed when this package is.
+As for the version numbers in parentheses,
+we got those by running `packageVersion("readr")` and similar commands inside RStudio
+and then rounding off.
+(See Section \@ref(branches-tag) for a discussion of version numbering.)
+
+All right: are we done now?
+No, we are not:
+
+```text
+❯ checking R code for possible problems ... NOTE
+  tally_words: no visible binding for global variable ‘word’
+  Undefined global functions or variables:
+    word
+```
+
+This is annoying but understandable.
+When the package builder is checking our code,
+it has no idea what columns are going to be in our data frames,
+so it has no way to know if `group_by(word)` will cause a problem.
+However,
+this is just a `NOTE`, not an `ERROR`,
+so we can try running "Build...Install and Restart"
+to build our package,
+re-start our R session (so that memory is clean),
+and load our newly-created package.
+It works!
+
+But there's still that warning.
+To fix it,
+we add this to our roxygen2 comment block:
+
+```{r eval=FALSE}
+#' @importFrom rlang .data
+```
+
+and then modify the calls that use naked column names to be:
+
+```{r eval=FALSE}
+tally_words <- function(all_words) {
+  dplyr::tibble(word = all_words) %>%
+    dplyr::group_by(.data$word) %>%
+    dplyr::tally()
+}
+```
+
+`.data` is a pronoun that allows us to be explicit when we refer to an object inside the data.
+Adding this---i.e.,
+being explicity that `word` is a column of `.data` rather than an undefined variable---finally
+gives us a clean build.
+
+### Exercise: Document dependencies
+
+1.  Modify the functions in the CO2 package to use `package::function` name for everything.
+2.  Modify the `DESCRIPTION` file to document the package's dependencies.
+
+## How can I add data to a package? {#rse-r-package-data}
+
+But we still need to add our cleaned-up data to our package
+and document the package as a whole.
+Working at the console,
+let's create a very small data set:
+
+```{r eval=FALSE}
+words <- c("some", "words", "appear", "some", "times")
+small_data <- tally_words(words)
+small_data
+```
+```
+# A tibble: 4 x 2
+  word       n
+  <chr>  <int>
+1 appear     1
+2 some       2
+3 times      1
+4 words      1
+```
+
+Next,
+we load the `usethis` package with `library(usethis)`
+and called `usethis::use_data(small_data)`
```

Comment:
Don't need to load `library(usethis)` when using `usethis::` notation, since `::` calls the function directly from usethis without needing `library()`.

Reply:
:+1: